apps/api/src/app/workflows-v2/usecases/upsert-workflow/upsert-workflow.usecase.ts

```diff
@@ -0,0 +1,281 @@
+  }
+
+  private async upsertControlValuesForSingleStep(
+    persistedStep: NotificationStepEntity,
+    command: UpsertWorkflowCommand,
+    persistedWorkflow: NotificationTemplateEntity
+  ): Promise<ControlValuesEntity | undefined> {
+    const stepDatabaseId = persistedStep._templateId;
+    const stepExternalId = persistedStep.name;
+    if (!stepDatabaseId && !stepExternalId) {
+      throw new StepUpsertMechanismFailedMissingIdException(stepDatabaseId, stepExternalId, persistedStep);
+    }
+    const stepInDto = command.workflowDto?.steps.find((commandStepItem) => commandStepItem.name === persistedStep.name);
+    if (!stepInDto) {
+      // TODO: should delete the values from the database?  or just ignore?
+      return;
+    }
+
+    const upsertControlValuesCommand = buildUpsertControlValuesCommand(
+      command,
+      persistedStep,
+      persistedWorkflow,
+      stepInDto
+    );
+
+    return await this.upsertControlValuesUseCase.execute(upsertControlValuesCommand);
+  }
+
+  private async upsertPreference(command: UpsertWorkflowCommand, workflow: NotificationTemplateEntity) {
+    let createdPreferences: PreferencesEntity | undefined;
+    if (command.workflowDto.preferences) {
+      createdPreferences = await this.upsertPreferencesUsecase.upsertWorkflowPreferences(
+        UpsertWorkflowPreferencesCommand.create({
+          environmentId: workflow._environmentId,
+          organizationId: workflow._organizationId,
+          templateId: workflow._id,
+          preferences: command.workflowDto.preferences,
+        })
+      );
+    }
+
+    return createdPreferences;
+  }
+
+  private async createOrUpdateWorkflow(
+    existingWorkflow: NotificationTemplateEntity | null | undefined,
+    command: UpsertWorkflowCommand
+  ): Promise<NotificationTemplateEntity> {
+    if (existingWorkflow) {
+      return await this.updateWorkflowUsecase.execute(
+        UpdateWorkflowCommand.create(this.convertCreateToUpdateCommand(command, existingWorkflow))
+      );
+    }
+
+    return await this.createWorkflowGenericUsecase.execute(
+      CreateWorkflowCommand.create(await this.buildCreateWorkflowGenericCommand(command))
+    );
+  }
+
+  private async validateEnvironment(command: UpsertWorkflowCommand) {
+    const environment = await this.environmentRepository.findOne({ _id: command.user.environmentId });
+
+    if (!environment) {
+      throw new BadRequestException('Environment not found');
+    }
+  }
+
+  private async buildCreateWorkflowGenericCommand(command: UpsertWorkflowCommand): Promise<CreateWorkflowCommand> {
+    const { user } = command;
+    const { workflowDto } = command;
+    const isWorkflowActive = command.workflowDto?.active ?? true;
+    const notificationGroupId = await this.getNotificationGroup(command.user.environmentId);
+
+    if (!notificationGroupId) {
+      throw new BadRequestException('Notification group not found');
+    }
+
+    return {
+      draft: !isWorkflowActive,
```

Comment:
good point.

`draft` is an old parameter for autocompletion that was never actually implemented. I think we can make it optional since it's already set as `{ default: true }` in the Mongoose schema.